### PR TITLE
Search action

### PR DESCRIPTION
**Pull request by frostealth**
Added `strongsquirrel\actions\SearchAction`.

**Usage**:
Controller `SiteController.php`

``` php
public function actions()
{
    return [
        'search' => [
            'class' => SearchAction::className(),
            'modelClass' => MySearchModel::className(),
            'httpMethod' => SearchAction::HTTP_METHOD_GET,
        ],
    ];
}
```

Model `MySearchModel.php`

``` php
class MySearchModel extends Model
{
    // ...

    public function search()
    {
        $query = MyModel::find();
        if ($this->validate()) {
            $query->filterWhere($this->getAttributes());
        }

        return new ActiveDataProvider(['query' => $query]);
    }
}
```

View `search.php`

``` php
echo GridView::widget([
    'dataProvider' => $dataProvider, // instance of [[ActiveDataProvider]]
    'filterModel' => $filterModel, // instance of [[MySearchModel]]
    // ...
]);
```

### CODE REVIEW

**Comment by chuprik** [comment]
Could you add author phpdoc in each classes?

File: src/SearchAction.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * @var string the name of the view action.
+     */
+    public $view = 'search';
+
+    /**
+     * @var string
+     */
+    public $httpMethod = self::HTTP_METHOD_POST;
```
Comment (chuprik) [inline src/SearchAction.php]: Why POST is priority?

**Comment by frostealth** [comment]
Only classes or methods too?

File: src/SearchAction.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+     * The method should return an instance of [[DataProviderInterface]].
+     *
+     * @var string
+     */
+    public $searchMethod = 'search';
+
+    /**
+     * @var array
+     */
+    public $searchOptions = [];
```
Comment (chuprik) [inline src/SearchAction.php:45]: When I can use it? 1-2 use cases please.

File: src/SearchAction.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * @return array [$dataProvider, $filterModel]
+     */
+    protected function prepare()
+    {
+        /** @var Model $model */
+        $model = new $this->modelClass(['scenario' => $this->scenario]);
+
+        $data = call_user_func([\Yii::$app->request, $this->httpMethod]);
```
Comment (chuprik) [inline src/SearchAction.php]: Magic :)

Comment (frostealth) [inline src/SearchAction.php]: Got any suggestions?

File: src/SearchAction.php (hunk 2 above)
Comment (frostealth) [inline src/SearchAction.php:45]: In the `search` method of the `SearchModel`.

``` php
public function search(array $options)
{
    // ...
    if (!empty($options['inactive'])) {
        $query->inactive();
    }
    // ...
}
```

File: src/SearchAction.php (hunk 1 above)
Comment (chuprik) [inline src/SearchAction.php]: I'm sure that $httpMethod should allow only proper verb. For example: I send a GET request but $httpMethod is POST and I think that action should throw exception in such situation.

File: src/SearchAction.php (hunk 3 above)
Comment (chuprik) [inline src/SearchAction.php]: look this https://github.com/StrongSquirrel/yii2-actions/pull/4/files#r44055992

File: src/SearchAction.php (hunk 1 above)
Comment (frostealth) [inline src/SearchAction.php]: No. This defines where to get the data for the model.

Comment (frostealth) [inline src/SearchAction.php]: Maybe rename it to `$formMethod` or `$searchMethod`?

Comment (chuprik) [inline src/SearchAction.php]: Then we need to write about source for model. To rename the variable would be better option. And replace `$data = call_user_func([\Yii::$app->request, $this->httpMethod]);` to simple if/else.

Comment (chuprik) [inline src/SearchAction.php]: > Maybe rename it to $formMethod or $searchMethod?

yes, I think it does makes sense.

**Comment by chuprik** [comment]
> Only classes or methods too?

all classes, I want see you as author.

**Comment by frostealth** [comment]
Done

File: src/SearchAction.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        $request = \Yii::$app->request;
+        switch ($this->formMethod) {
+            case self::FORM_METHOD_GET:
+                $data = $request->get();
+                break;
+            case self::FORM_METHOD_POST:
+                $data = $request->post();
+                break;
+            default:
+                $data = null;
```
Comment (chuprik) [inline src/SearchAction.php]: Maybe exception? Because using PUT and DELETE for search is very-very bad way.

**Comment by frostealth** [comment]
All right?

**Comment by chuprik** [comment]
yeah. thank you!

**Pull request merged**